File: apis/core/src/eave/internal/openai.py
import enum
import logging
from dataclasses import asdict, dataclass
from typing import Any, Optional, cast

import openai as openai_sdk
import openai.openai_object
import tiktoken

import eave.internal.settings
import eave.internal.util

tokencoding = tiktoken.get_encoding("gpt2")

PROMPT_PREFIX = (
    "You are Eave, an AI documentation expert. "
    "Your job is to write, find, and organize robust, detailed documentation of this organization's information, decisions, projects, and procedures. "
    "You are responsible for the quality and integrity of this organization's documentation."
)


class OpenAIModel(str, enum.Enum):
    ADA = "text-ada-001"
    ADA_EMBEDDING = "text-embedding-ada-002"
    CURIE = "text-curie-001"
    BABBAGE = "text-babbage-001"
    DAVINCI = "text-davinci-003"


MAX_TOKENS = 4096  # 2048 if using older models


@dataclass
class CompletionParameters:
    prompt: str = ""
    best_of: Optional[int] = None
    n: Optional[int] = None
    frequency_penalty: Optional[float] = None
    presence_penalty: Optional[float] = None
    temperature: Optional[float] = None

    def compile(self) -> eave.internal.util.JsonObject:
        params = dict[str, Any]()
        params["prompt"] = self.prompt
        params["model"] = OpenAIModel.DAVINCI
        params["max_tokens"] = MAX_TOKENS - len(tokencoding.encode(self.prompt))

        if self.best_of is not None:
            params["best_of"] = self.best_of
        if self.n is not None:
            params["n"] = self.n
        if self.frequency_penalty is not None:
            params["frequency_penalty"] = self.frequency_penalty
        if self.presence_penalty is not None:
            params["presence_penalty"] = self.presence_penalty
        if self.temperature is not None:
            params["temperature"] = self.temperature
        return params


async def summarize(params: CompletionParameters) -> Optional[str]:
    """
    https://beta.openai.com/docs/api-reference/completions/create
    """

    # Set this here so that the openai key is pulled from GCP Secrets Manager lazily
    if openai_sdk.api_key is None:
        openai_sdk.api_key = eave.internal.settings.APP_SETTINGS.eave_openai_api_key

    response = await openai_sdk.Completion.acreate(**params.compile())
    response = cast(openai.openai_object.OpenAIObject, response)
    candidates = [c for c in response.choices if c["finish_reason"] == "stop"]

    if len(candidates) < 1:
        logging.warn("No valid choices from openAI")
        return None

    answer = str(candidates[0].text).strip()
    return answer


# def get_embedding(text, model=OpenAIModel.ADA_EMBEDDING):
#     text = text.replace("\n", " ")
#     return openai.Embedding.create(input=[text], model=model)["data"][0]["embedding"]


# df = pandas.read_csv("output/embedded_1k_reviews.csv")
# df["ada_embedding"] = df.ada_embedding.apply(eval).apply(numpy.array)
# df["ada_embedding"] = df.combined.apply(lambda x: get_embedding(x, model=OpenAIModel.ADA_EMBEDDING))
# df.to_csv("output/embedded_1k_reviews.csv", index=False)

File: libs/eave-stdlib-ts/src/openai.ts
import { v4 as uuidv4 } from 'uuid';
import { Configuration, OpenAIApi, CreateChatCompletionRequest, ChatCompletionRequestMessageRoleEnum } from 'openai';
import { encoding_for_model } from 'tiktoken';
import { sharedConfig } from './config.js';
import eaveLogger from './logging.js';
import { CtxArg } from './requests.js';

// eslint-disable-next-line operator-linebreak
export const PROMPT_PREFIX =
  'You are Eave, a documentation expert. '
  + "Your job is to write, find, and organize robust, detailed documentation of this organization's information, decisions, projects, and procedures. "
  + "You are responsible for the quality and integrity of this organization's documentation.";

export enum OpenAIModel {
  GPT_35_TURBO = 'gpt-3.5-turbo',
  GPT_35_TURBO_16K = 'gpt-3.5-turbo-16k',
  GPT4 = 'gpt-4',
  // GPT4_32K = 'gpt-4-32k',
}

function modelFromString(v: string): OpenAIModel | undefined {
  switch (v) {
    case OpenAIModel.GPT_35_TURBO: return OpenAIModel.GPT_35_TURBO;
    case OpenAIModel.GPT_35_TURBO_16K: return OpenAIModel.GPT_35_TURBO_16K;
    case OpenAIModel.GPT4: return OpenAIModel.GPT4;
    default: return undefined;
    // case OpenAIModel.GPT4_32K: return OpenAIModel.GPT4_32K;
  }
}

type MaxTokens = {
  [OpenAIModel.GPT_35_TURBO]: number;
  [OpenAIModel.GPT_35_TURBO_16K]: number;
  [OpenAIModel.GPT4]: number;
  // [OpenAIModel.GPT4_32K]: number;
}

const MAX_TOKENS: MaxTokens = {
  [OpenAIModel.GPT_35_TURBO]: 4096,
  [OpenAIModel.GPT_35_TURBO_16K]: 16384,
  [OpenAIModel.GPT4]: 8192,
  // [OpenAIModel.GPT4_32K]: 32768,
};

export default class OpenAIClient {
  client: OpenAIApi;

  static tokenCount(data: string, model: OpenAIModel): number {
    const encoder = encoding_for_model(model);
    const count = encoder.encode(data).length;
    encoder.free();
    return count;
  }

  static maxTokens(model: OpenAIModel): number {
    return MAX_TOKENS[model];
  }

  static async getAuthedClient(): Promise<OpenAIClient> {
    const apiKey = await sharedConfig.openaiApiKey;
    const apiOrg = await sharedConfig.openaiApiOrg;
    const configuration = new Configuration({ apiKey, organization: apiOrg });
    const openaiClient = new OpenAIApi(configuration);
    return new OpenAIClient(openaiClient);
  }

  constructor(client: OpenAIApi) {
    this.client = client;
  }

  /*
    https://beta.openai.com/docs/api-reference/completions/create
    baseTimeoutSeconds is multiplied by (2^n) for each attempt n
  */
  async createChatCompletion({ parameters, ctx, baseTimeoutSeconds = 30 }: CtxArg & {parameters: CreateChatCompletionRequest, baseTimeoutSeconds?: number}): Promise<string> {
    parameters.messages.unshift({ role: ChatCompletionRequestMessageRoleEnum.System, content: PROMPT_PREFIX });

    const model = modelFromString(parameters.model);
    if (!model) {
      throw new Error(`Unexpected model value ${parameters.model}`);
    }

    const logParams = {
      openai: <any>parameters,
      openaiRequestId: uuidv4(),
    };

    let text: string | undefined;

    const maxAttempts = 3;
    for (let i = 0; i < maxAttempts; i += 1) {
      const backoffMs = baseTimeoutSeconds * (2 ** i) * 1000;
      try {
        eaveLogger.debug('openai request', ctx, logParams);
        const completion = await this.client.createChatCompletion(parameters, { timeout: backoffMs }); // timeout in ms
        eaveLogger.debug('openai response', logParams, { openaiResponse: <any>completion.data }, ctx);
        text = completion.data.choices[0]?.message?.content;
        break;
      } catch (e: any) {
        // Network error?
        if (i + 1 < maxAttempts) {
          eaveLogger.warning(e, logParams, ctx);
          await new Promise((r) => { setTimeout(r, backoffMs); });
        } else {
          throw e;
        }
      }
    }

    if (text === undefined) {
      throw new Error('openai text response is undefined');
    }
    return text;
  }
}
